pkg/deploy/controller/deployerpod/controller_test.go

```diff
@@ -160,39 +166,149 @@ func TestHandle_podTerminatedFail(t *testing.T) {
 	controller := &DeployerPodController{
 		deploymentClient: &deploymentClientImpl{
 			getDeploymentFunc: func(namespace, name string) (*kapi.ReplicationController, error) {
-				config := deploytest.OkDeploymentConfig(1)
-				deployment, _ := deployutil.MakeDeployment(config, kapi.Codec)
-				deployment.Annotations[deployapi.DeploymentStatusAnnotation] = string(deployapi.DeploymentStatusRunning)
-				return deployment, nil
+				return failedDeployment, nil
 			},
 			updateDeploymentFunc: func(namespace string, deployment *kapi.ReplicationController) (*kapi.ReplicationController, error) {
-				updatedDeployment = deployment
+				if _, found := updatedDeployments[deployutil.DeploymentVersionFor(deployment)]; found {
+					t.Fatalf("unexpected multiple updates for the same deployment %d", deployutil.DeploymentVersionFor(deployment))
+				}
+				updatedDeployments[deployutil.DeploymentVersionFor(deployment)] = deployment
 				return deployment, nil
 			},
+			listDeploymentsForConfigFunc: func(namespace, configName string) (*kapi.ReplicationControllerList, error) {
+				return existingDeployments, nil
+			},
 		},
+		scaler: &scaler,
 	}
 
-	err := controller.Handle(terminatedPod())
+	type existing struct {
+		version         int
+		status          deployapi.DeploymentStatus
+		initialReplicas int
+		updatedReplicas int
+	}
 
-	if err != nil {
-		t.Fatalf("unexpected error: %v", err)
+	type scenario struct {
+		// this is the deployment that is passed to Handle
+		version int
+		// this is the target replicas for the deployment that failed
+		desiredReplicas int
+		// existing deployments also include the one being handled currently
+		existing []existing
 	}
 
-	if updatedDeployment == nil {
-		t.Fatalf("expected deployment update")
+	scenarios := []scenario{
+		// No previous deployments
+		{1, 3, []existing{
+			{1, deployapi.DeploymentStatusRunning, 3, 0},
+		}},
+		// Multiple existing deployments - none in complete state
+		{3, 2, []existing{
+			{1, deployapi.DeploymentStatusFailed, 2, 2},
+			{2, deployapi.DeploymentStatusFailed, 0, 0},
+			{3, deployapi.DeploymentStatusRunning, 2, 0},
+		}},
+		// Failed deployment is already at 0 replicas
+		{3, 2, []existing{
+			{1, deployapi.DeploymentStatusFailed, 2, 2},
+			{2, deployapi.DeploymentStatusFailed, 0, 0},
+			{3, deployapi.DeploymentStatusRunning, 0, 0},
+		}},
+		// Multiple existing completed deployments
+		{4, 2, []existing{
+			{3, deployapi.DeploymentStatusComplete, 0, 2},
+			{2, deployapi.DeploymentStatusComplete, 0, 0},
+			{4, deployapi.DeploymentStatusRunning, 1, 0},
+			{1, deployapi.DeploymentStatusFailed, 0, 0},
+		}},
+		// A deployment already exists after the current failed deployment
+		{4, 2, []existing{
+			{3, deployapi.DeploymentStatusComplete, 1, 1},
+			{2, deployapi.DeploymentStatusComplete, 0, 0},
+			{4, deployapi.DeploymentStatusRunning, 2, 0},
+			{5, deployapi.DeploymentStatusNew, 0, 0},
+			{1, deployapi.DeploymentStatusFailed, 0, 0},
```

Comment:
Are these out of order because you are verifying sort?  If so, add a comment to that effect